crypto/bn/asm/rsaz-3k-avx512.pl

```diff
@@ -0,0 +1,885 @@
+$code.=<<___ if ($win64);
+    vmovdqa64   `0*16`(%rax),%xmm6
+    vmovdqa64   `1*16`(%rax),%xmm7
+    vmovdqa64   `2*16`(%rax),%xmm8
+    vmovdqa64   `3*16`(%rax),%xmm9
+    vmovdqa64   `4*16`(%rax),%xmm10
+    vmovdqa64   `5*16`(%rax),%xmm11
+    vmovdqa64   `6*16`(%rax),%xmm12
+    vmovdqa64   `7*16`(%rax),%xmm13
+    vmovdqa64   `8*16`(%rax),%xmm14
+    vmovdqa64   `9*16`(%rax),%xmm15
+    lea     168(%rsp),%rax
+___
+$code.=<<___;
+    mov  0(%rax),%r15
+.cfi_restore    %r15
+    mov  8(%rax),%r14
+.cfi_restore    %r14
+    mov  16(%rax),%r13
+.cfi_restore    %r13
+    mov  24(%rax),%r12
+.cfi_restore    %r12
+    mov  32(%rax),%rbp
+.cfi_restore    %rbp
+    mov  40(%rax),%rbx
+.cfi_restore    %rbx
+    lea  48(%rax),%rsp
+.cfi_def_cfa    %rsp,8
+.Lrsaz_amm52x30_x2_ifma256_epilogue:
+    ret
+.cfi_endproc
+.size   RSAZ_amm52x30_x2_ifma256, .-RSAZ_amm52x30_x2_ifma256
+___
+}
+
+###############################################################################
+# Constant time extraction from the precomputed table of powers base^i, where
+#    i = 0..2^EXP_WIN_SIZE-1
+#
+# The input |red_table| contains precomputations for two independent base values,
+# so the |tbl_idx| indicates for which base shall we extract the value.
+# |red_table_idx| is a power index.
+#
+# Extracted value (output) is 30 + 2 digit number in 2^52 radix.
+# (2 high QW is zero padding)
+#
+# void extract_multiplier_2x30_win5(BN_ULONG *red_Y,
+#                                   const BN_ULONG red_table[1 << EXP_WIN_SIZE][2][32],
+#                                   int red_table_idx,
+#                                   int tbl_idx);           # 0 or 1
+#
+# EXP_WIN_SIZE = 5
+###############################################################################
+{
+# input parameters
+my ($out,$red_tbl,$red_tbl_idx,$tbl_idx)=$win64 ? ("%rcx","%rdx","%r8", "%r9") :  # Win64 order
+                                                  ("%rdi","%rsi","%rdx","%rcx");  # Unix order
+
+my ($t0,$t1,$t2,$t3,$t4,$t5) = map("%ymm$_", (0..5));
+my ($t6,$t7) = map("%ymm$_", (16..17));
+my $t0xmm = $t0 =~ s/%y/%x/r;
+my ($tmp0,$tmp1,$tmp2,$tmp3,$tmp4,$tmp5,$tmp6,$tmp7) = map("%ymm$_", (20..27));
+my ($cur_idx,$idx) = map("%ymm$_", (30..31));
+
+$code.=<<___;
+.text
+
+.align 32
+.globl  extract_multiplier_2x30_win5
+.type   extract_multiplier_2x30_win5,\@abi-omnipotent
+extract_multiplier_2x30_win5:
+.cfi_startproc
+    endbranch
+    movq    $tbl_idx, %rax
+    imulq   \$`32*8`, %rax
+    addq    %rax, $red_tbl
+
+    vpbroadcastq    $red_tbl_idx, $idx
+    leaq   `(1<<5)*2*32*8`($red_tbl), %rax  # holds end of the tbl
+
+    vpxor   $t0xmm, $t0xmm, $t0xmm
+    vmovdqa64   $t0, $t7
+    vmovdqa64   $t0, $t6
+    vmovdqa64   $t0, $t5
+    vmovdqa64   $t0, $t4
+    vmovdqa64   $t0, $t3
+    vmovdqa64   $t0, $t2
+    vmovdqa64   $t0, $t1
+    vmovdqa64   $t0, $cur_idx
+
+.align 32
+.Lloop:
+    vpcmpq  \$0, $cur_idx, $idx, %k1             # mask of (idx == cur_idx)
+    addq    \$`2*32*8`, $red_tbl
```

Comment:
similar to extract_multiplier_2x20_win5, why not delay increment, then can remove '-2*32*8' from vmovdqu64 instructions which follow, just for code readability

Reply:
comment no longer relevant with new implementation